dm/dm/master/openapi_controller.go

```diff
@@ -49,20 +58,65 @@ func (s *Server) getSourceStatusListFromWorker(ctx context.Context, sourceName s
 				Stage:              relayStatus.Stage.String(),
 			}
 		}
+		// add error if some error happen
+		if workerStatus.SourceStatus.Result != nil && len(workerStatus.SourceStatus.Result.Errors) > 0 {
+			var errorMsgs string
+			for _, err := range workerStatus.SourceStatus.Result.Errors {
+				errorMsgs += fmt.Sprintf("%s\n", err.Message)
+			}
+			sourceStatus.ErrorMsg = &errorMsgs
+		}
 		sourceStatusList[i] = sourceStatus
 	}
 	return sourceStatusList, nil
 }
 
-// nolint:unparam
-func (s *Server) createSource(ctx context.Context, cfg *config.SourceConfig) error {
-	return s.scheduler.AddSourceCfg(cfg)
+func (s *Server) createSource(ctx context.Context, req openapi.CreateSourceRequest) (*openapi.Source, error) {
+	cfg := config.OpenAPISourceToSourceCfg(req.Source)
+	if err := checkAndAdjustSourceConfigFunc(ctx, cfg); err != nil {
+		return nil, err
+	}
+
+	var err error
+	if req.WorkerName == nil {
+		err = s.scheduler.AddSourceCfg(cfg)
+	} else {
+		err = s.scheduler.AddSourceCfgWithWorker(cfg, *req.WorkerName)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if cfg.EnableRelay {
+		return &req.Source, s.enableRelay(ctx, req.Source.SourceName, openapi.EnableRelayRequest{})
+	}
+	return &req.Source, nil
 }
 
-// nolint:unparam,unused
-func (s *Server) updateSource(ctx context.Context, cfg *config.SourceConfig) error {
-	// TODO(ehco) no caller now , will implement later
-	return nil
+func (s *Server) updateSource(ctx context.Context, sourceName string, req openapi.UpdateSourceRequest) (*openapi.Source, error) {
+	// TODO: support dynamic updates
+	oldCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if oldCfg == nil {
+		return nil, terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	newCfg := config.OpenAPISourceToSourceCfg(req.Source)
+	if err := checkAndAdjustSourceConfigFunc(ctx, newCfg); err != nil {
+		return nil, err
+	}
+	if err := s.scheduler.UpdateSourceCfg(newCfg); err != nil {
+		return nil, err
+	}
+	// when enable filed updated, we need operate task on this source
```

Comment:
For the moment, relay is not affected, the state of relay is not entirely determined by source.

This will need to be optimized when the relay state is standardized